client/container_stats.go

```diff
@@ -6,61 +6,75 @@ import (
+	// cancelled.
+	Stream bool
+
+	// IncludePreviousSample asks the daemon to  collect a prior sample to populate the
+	// [container.StatsResponse.PreRead] and [container.StatsResponse.PreCPUStats]
+	// fields.
+	//
+	// It set, the daemon collects two samples at a one-second interval before
+	// returning the result. The first sample populates the PreCPUStats (“previous
+	// CPU”) field, allowing delta calculations for CPU usage. If false, only
+	// a single sample is taken and returned immediately, leaving PreRead and
+	// PreCPUStats empty.
+	//
+	// This option has no effect if Stream is enabled. If Stream is enabled,
+	// [container.StatsResponse.PreCPUStats] is never populated for the first
+	// record.
+	IncludePreviousSample bool
+}
+
+// ContainerStatsResult holds the result from [Client.ContainerStats].
 //
-// The OSType field is set to the server's platform to allow
-// platform-specific handling of the response.
+// It wraps an [io.ReadCloser] that provides one or more [container.StatsResponse]
+// objects for a container, as produced by the "GET /containers/{id}/stats" endpoint.
+// If streaming is disabled, the stream contains a single record.
 //
-// TODO(thaJeztah): remove this wrapper, and make OSType part of [github.com/moby/moby/api/types/container.StatsResponse].
-type StatsResponseReader struct {
-	Body   io.ReadCloser `json:"body"`
-	OSType string        `json:"ostype"`
+// The OSType field reports the daemon's operating system, allowing platform-specific
+// handling of the response.
+type ContainerStatsResult struct {
+	Body   io.ReadCloser
+	OSType string // TODO(thaJeztah): consider moving OSType into [container.StatsResponse].
 }
 
-// ContainerStats returns near realtime stats for a given container.
-// It's up to the caller to close the [io.ReadCloser] returned.
-func (cli *Client) ContainerStats(ctx context.Context, containerID string, stream bool) (StatsResponseReader, error) {
+// ContainerStats retrieves live resource usage statistics for the specified
+// container. The caller must close the [io.ReadCloser] in the returned result
+// to release associated resources.
+//
+// If the stream ends or the context is canceled, the underlying [io.ReadCloser]
+// is automatically closed.
+func (cli *Client) ContainerStats(ctx context.Context, containerID string, options ContainerStatsOptions) (ContainerStatsResult, error) {
 	containerID, err := trimID("container", containerID)
 	if err != nil {
-		return StatsResponseReader{}, err
+		return ContainerStatsResult{}, err
 	}
 
 	query := url.Values{}
-	query.Set("stream", "0")
-	if stream {
-		query.Set("stream", "1")
-	}
-
-	resp, err := cli.get(ctx, "/containers/"+containerID+"/stats", query, nil)
-	if err != nil {
-		return StatsResponseReader{}, err
+	if options.Stream {
+		query.Set("stream", "true")
+	} else {
+		// Note: daemons may return an error if both set: "cannot have stream=true and one-shot=true"
```

Comment:
Shouldn't we still pass it to the daemon so that it can return an error (or actually handle it when it supports it)?

Otherwise passing both `Stream = true and IncludePreviousSample = true` will still be silently accepted even though it's not a valid setting.